src/inmanta/server/services/compilerservice.py

```diff
@@ -774,15 +781,44 @@ async def _sub_notify() -> None:
 
     async def _recover(self) -> None:
         """Restart runs after server restart"""
+        # patch up in-memory state
         async with self._queue_count_cache_lock:
             self._queue_count_cache = await data.Compile.get_total_length_of_all_compile_queues(exclude_started_compiles=False)
-        # one run per env max to get started
-        runs = await data.Compile.get_next_run_all()
-        for run in runs:
-            await self._queue(run)
-        unhandled = await data.Compile.get_unhandled_compiles()
-        for u in unhandled:
-            await self._notify_listeners(u)
+
+        # we run the recovery and start in the background to prevent deadlocks with sync handlers
+        async def sub_recovery() -> None:
+            unhandled = await data.Compile.get_unhandled_compiles()
+            for u in unhandled:
+                await self._notify_listeners(u)
+            # one run per env max to get started
+            runs = await data.Compile.get_next_run_all()
+
+            # prevent races by holding global lock
+            async with self._global_lock:
+                # We are ready to start!
+                self.fully_ready = True
+                for run in runs:
+                    await self.__queue_no_lock(run.environment, run)
+
+        self.add_background_task(sub_recovery())
+
+    async def __queue_no_lock(self, environment: uuid.UUID, compile: data.Compile) -> None:
```

Comment:
Is there a reason why `_queue()` doesn't call into this method? Now this code is duplicates across two methods.